flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java

```diff
@@ -0,0 +1,344 @@
+		TypeSerializer<T> serializer,
+		Properties kafkaProperties,
+		long pollTimeout,
+		MetricGroup subtaskMetricGroup,
+		MetricGroup consumerMetricGroup,
+		boolean useMetrics,
+		int producerParallelism) throws Exception {
+		super(
+			sourceContext,
+			assignedPartitionsWithInitialOffsets,
+			watermarksPeriodic,
+			watermarksPunctuated,
+			processingTimeProvider,
+			autoWatermarkInterval,
+			userCodeClassLoader,
+			consumerMetricGroup,
+			useMetrics);
+
+		this.deserializer = new KafkaShuffleElementDeserializer<>();
+		this.serializer = serializer;
+		this.handover = new Handover();
+		this.consumerThread = new KafkaConsumerThread(
+			LOG,
+			handover,
+			kafkaProperties,
+			unassignedPartitionsQueue,
+			getFetcherName() + " for " + taskNameWithSubtasks,
+			pollTimeout,
+			useMetrics,
+			consumerMetricGroup,
+			subtaskMetricGroup);
+		this.watermarkHandler = new WatermarkHandler(producerParallelism);
+	}
+
+	// ------------------------------------------------------------------------
+	//  Fetcher work methods
+	// ------------------------------------------------------------------------
+
+	@Override
+	public void runFetchLoop() throws Exception {
+		try {
+			final Handover handover = this.handover;
+
+			// kick off the actual Kafka consumer
+			consumerThread.start();
+
+			while (running) {
+				// this blocks until we get the next records
+				// it automatically re-throws exceptions encountered in the consumer thread
+				final ConsumerRecords<byte[], byte[]> records = handover.pollNext();
+
+				// get the records for each topic partition
+				for (KafkaTopicPartitionState<TopicPartition> partition : subscribedPartitionStates()) {
+					List<ConsumerRecord<byte[], byte[]>> partitionRecords =
+						records.records(partition.getKafkaPartitionHandle());
+
+					for (ConsumerRecord<byte[], byte[]> record : partitionRecords) {
+						final KafkaShuffleElement<T> element = deserializer.deserialize(serializer, record);
+
+						// TODO: do we need to check the end of stream if reaching the end watermark?
+
+						if (element.isRecord()) {
+							// timestamp is inherent from upstream
+							// If using ProcessTime, timestamp is going to be ignored (upstream does not include timestamp as well)
+							// If using IngestionTime, timestamp is going to be overwritten
+							// If using EventTime, timestamp is going to be used
+							synchronized (checkpointLock) {
+								KafkaShuffleRecord<T> elementAsRecord = element.asRecord();
+								sourceContext.collectWithTimestamp(
+									elementAsRecord.value,
+									elementAsRecord.timestamp == null ? record.timestamp() : elementAsRecord.timestamp);
+								partition.setOffset(record.offset());
+							}
+						} else if (element.isWatermark()) {
+							final KafkaShuffleWatermark watermark = element.asWatermark();
+							Optional<Watermark> newWatermark = watermarkHandler.checkAndGetNewWatermark(watermark);
+							newWatermark.ifPresent(sourceContext::emitWatermark);
```

Comment:
This is a rewrite of {@link AbstractFetcher#emitRecordWithTimestamp}

         if (record != null) {
			if (timestampWatermarkMode == NO_TIMESTAMPS_WATERMARKS) {
				// fast path logic, in case there are no watermarks generated in the fetcher

				// emit the record, using the checkpoint lock to guarantee
				// atomicity of record emission and offset state update
				synchronized (checkpointLock) {
					sourceContext.collectWithTimestamp(record, timestamp);
					partitionState.setOffset(offset);
				}
			} else if (timestampWatermarkMode == PERIODIC_WATERMARKS) {
				emitRecordWithTimestampAndPeriodicWatermark(record, partitionState, offset, timestamp);
			} else {
				emitRecordWithTimestampAndPunctuatedWatermark(record, partitionState, offset, timestamp);
			}
		} else {
			// if the record is null, simply just update the offset state for partition
			synchronized (checkpointLock) {
				partitionState.setOffset(offset);
			}
		}

Reply:
If you go through `emitRecordWithTimestampAndPunctuatedWatermark` 
for exmaple

records are emitted under checkpointLock
watermarks are emitted through updateMinPunctuatedWatermark(newWatermark), not within the lock.

This is understandable in the current case, since watermark is only a indicator, not strictly part of a `checkpoint`. That's why.

Reply:
Yes, you are not modifying state, so all good.